docs: add s3-email-blog (#251)
linkall-labs/docs: blog/2023-02-07-send-email-notification-on-AWS-S3-events/index.md

authors: Jay
tags: ['S3', 'Email', 'events']
---

# Send email notification on AWS S3 events

One of the most common scenarios when working with Amazon S3 is to perform specific actions whenever changes (e.g., file uploaded or file deleted) happen in the S3 buckets.

This article introduces an **open-source solution** to set email notifications on any S3 events with an **open-source event streaming platform** [Vanus](https://github.com/linkall-labs/vanus).

<!--truncate-->

## Table of Contents
- [About AWS S3](#about-aws-s3)
    - [What is AWS S3](#what-is-aws-s3)
    - [What is AWS events](#what-is-aws-events)
    - [Why we need AWS events](#why-we-need-aws-events)
- [Pre-requisite](#pre-requisite)
- [How to Send email notification on AWS S3 events](#how-to-send-email-notification-on-aws-s3-events)
    - [Step 1: Deploy Vanus on the Playground](#step-1-deploy-vanus-on-the-playground)
    - [Step 2: Deploy the S3 Source connector](#step-2-deploy-the-s3-source-connector)
    - [Step 3: Deploy the Email Sink connector](#step-3-deploy-the-email-sink-connector)
- [Check out the result](#check-out-the-result)  
- [Conclusion](#conclusion)

## About AWS S3

> ## What is AWS S3

Reviewer: ```suggestion
### What is AWS S3
```

Amazon Simple Storage Service (Amazon S3) is an object storage service that offers industry-leading scalability, data availability, security, and performance. This means customers of all sizes and industries can use it to store and protect any amount of data for a range of use cases, such as websites, mobile applications, backup and restore, archive, enterprise applications, IoT devices, and big data analytics.

With Amazon S3, you can store your data in a highly durable manner, as it automatically stores multiple copies of data across multiple devices in multiple facilities, as well as automatically detecting and repairing any lost redundancy. Additionally, it integrates with other AWS services, making it easy to use as the foundation of your application's data architecture.

> ## What is AWS events

Reviewer: ```suggestion
### What is AWS events
```

Amazon S3 events refer to changes or updates to objects stored in an Amazon S3 bucket. S3 events can trigger a variety of actions, such as sending a notification, starting an AWS Lambda function, or triggering an Amazon SNS message.

Currently, Amazon S3 can publish notifications for the following events:

- New object created events
- Object removal events
- Restore object events
- Reduced Redundancy Storage (RRS) object lost events
- Replication events
- S3 Lifecycle expiration events
- S3 Lifecycle transition events
- S3 Intelligent-Tiering automatic archival events
- Object tagging events
- Object ACL PUT events

> ## Why we need AWS events

Reviewer: ```suggestion
### Why we need AWS events
```

vi sink-email.yaml
```
**NOTE**: Remember to replace values of `default`, `account`, and `password` with yours. The password should be a 16 character password generated
in previous step.

![img.png](img/img_4.png)

- Run the Email Sink.

```shell
kubectl apply -f sink-email.yaml
```

## Check out the result

Before checking out the result, let's recap what we've done in previous steps.

Step 1: We created a basic store component of Vanus - `EventBus` called `s3-email-scenario`.

Step 2: We deployed a S3 Source Connector to receive S3 events.

Step 3: We deployed an Email Sink Connector waiting to send emails to some contacts.

You may notice that we didn't specify the contact email address in Email Sink. So how the sink connector can know who it should send emails to ?
The Email Sink also doesn't know what content it should send.

The missing **key-point** here is a `subscription`. It can be used to grab data from specific `EventBus` and send data to sink targets (in our case the sink is Email Sink Connector).

Also, in a `subscription`, you can create `filters` to select data which meet your requirements, and create `transformers` to

> convert original CloudEvents into any format you need, before them sent to target.

Reviewer: ```suggestion
convert original CloudEvents into any format you need, before before sending them to the target.
```

Now let's create a subscription for this tutorial.

- Create a subscription

```shell
vsctl subscription create --name s3-email-subscription \
  --eventbus s3-email-scenario \
  --sink 'http://sink-email:8080' \
  --transformer '{
      "define": {
        "subject" : "$.subject",
        "bucket": "$.data.bucket.name"
      },
      "pipeline":[
        {"command":["create","$.xvemailrecipients","contact@linkall.com"]},
        {"command":["create","$.xvemailsubject","Greetings from Vanus-S3-Email-Scenario"]}
      ],
      "template": "Hello, a new file <subject> has been uploaded to your bucket <bucket>."
    }'
```

We created a `subscription` called `s3-email-subscription`, and specified that it will send data to our Email Sink connector.

We also created a transformer which did two things:

1. Added two [CloudEvents attributes](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#context-attributes) `xvemailrecipients` to specify the email recipient and `xvemailsubject` to specify the email subject.

> 2. Assigned values from the original event to two defined variables `subject` and `bucket`.And use them to generate the content which will be sent to email contacts.

Reviewer: ```suggestion
2. Assigned values from the original event to two defined variables `subject` and `bucket`.And use them to generate the content which will be sent to our email recipient.
```

> Now, we can upload a file to see if an email notification can be sent.

Reviewer: ```suggestion
Now, we can upload a file to see if an email notification will be sent correctly.
```
